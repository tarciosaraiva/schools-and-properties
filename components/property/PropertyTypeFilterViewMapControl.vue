<template>
  <div class="layer-container">
    <button title="Show property type filter control" @click="toggleControl">
      <font-awesome-icon icon="home" />
    </button>
    <button v-show="expanded" title="Query listings with current property types" @click="handleChange">
      <font-awesome-icon icon="check" size="lg" :style="{ color: 'rgba(39, 174, 96)' }" />
    </button>
    <ul class="layer-options" :class="collapsed ? '' : 'expanded'">
      <li>
        <label for="p_house_type">
          <input
            id="p_house_type"
            v-model="propertyTypes"
            type="checkbox"
            value="House"
          /> House
        </label>
      </li>
      <li>
        <label for="p_townhouse_type">
          <input
            id="p_townhouse_type"
            v-model="propertyTypes"
            type="checkbox"
            value="Townhouse"
          /> Townhouse
        </label>
      </li>
      <li>
        <label for="p_unit_type">
          <input
            id="p_unit_type"
            v-model="propertyTypes"
            type="checkbox"
            value="ApartmentUnitFlat"
          /> Unit
        </label>
      </li>
    </ul>
  </div>
</template>

<script lang="ts">
import Vue from 'vue'
import { mapGetters, mapActions } from 'vuex'

export default Vue.extend({
  data () {
    return {
      collapsed: true
    }
  },
  computed: {
    ...mapGetters(['propertiesFilter']),
    expanded () { return !this.collapsed },
    propertyTypes: {
      get() {
        return this.propertiesFilter.types
      },
      set(value) {
        this.$store.commit('UPDATE_PROPERTY_FILTER_TYPE', value)
      },
    }
  },
  mounted() {
    document.addEventListener('click', this.handleClickOutside)
  },
  destroyed() {
    document.removeEventListener('click', this.handleClickOutside)
  },
  methods: {
    ...mapActions(['loadListings']),
    handleClickOutside(event: any) {
      if (!this.$el.contains(event.target)) {
        this.collapsed = true;
      }
    },
    toggleControl () {
      this.collapsed = !this.collapsed
    },
    handleChange() {
      this.collapsed = true
      this.loadListings()
    }
  }
})
</script>
