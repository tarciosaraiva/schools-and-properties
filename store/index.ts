import { GetterTree, ActionTree, MutationTree } from 'vuex'

interface ApplicationFilter {
  properties: PropertiesFilter,
  schools: SchoolsFilter
}

interface PropertiesFilter {
  types: string[],
  rooms: number,
  bathrooms: number,
  carSpaces: number,
  maxPrice: number
}

export interface SchoolsFilter {
  educationSector: string,
  zone: string,
  primary: SchoolLevelRating,
  secondary: SchoolLevelRating,
}

interface SchoolLevelRating {
  plot: boolean,
  rating: number,
  englishRating: number,
  mathsRating: number
}

export interface PropertyListing {
  id: number,
  price: string,
  area: string
  bathrooms: number,
  bedrooms: number,
  carspaces: number,
  displayableAddress: string,
  features: string[],
  landArea: number,
  latitude: number,
  longitude: number,
  postcode: string,
  propertyType: string
  region: string,
  state: string,
  street: string,
  streetNumber: string,
  suburb: string,
  unitNumber: string,
  media: string[],
  listingSlug: string
}

export interface School {
  schoolName: string,
  educationSector: string,
  schoolType: string,
  phoneNumber: string,
  primaryOverallScore: string,
  primaryEnglishScore: string,
  primaryMathsScore: string,
  primaryEnrolments: string,
  secondaryOverallScore: string,
  secondaryEnglishScore: string,
  secondaryMathsScore: string,
  secondaryEnrolments: string
}

export const state = () => ({
  bounds: {
    ne: { lat: 0, lon: 0 },
    nw: { lat: 0, lon: 0 },
    se: { lat: 0, lon: 0 },
    sw: { lat: 0, lon: 0 },
  },
  geocodeSuggestions: [] as any,
  list: [] as PropertyListing[],
  filter: {
    properties: {
      types: ['House', 'Townhouse'],
      rooms: 3,
      bathrooms: 2,
      carSpaces: 1,
      maxPrice: 900000
    },
    schools: {
      educationSector: 'all',
      zone: 'none',
      primary: {
        plot: true,
        rating: 95,
        englishRating: 4,
        mathsRating: 4
      },
      secondary: {
        plot: true,
        rating: 95,
        englishRating: 4,
        mathsRating: 4
      }
    }
  } as ApplicationFilter
})

export type RootState = ReturnType<typeof state>

export const getters: GetterTree<RootState, RootState> = {
  listings: state => state.list,
  schoolsFilter: state => state.filter.schools,
  propertiesFilter: state => state.filter.properties,
  locations: state => state.geocodeSuggestions,
  listingPois: state => state.list.map(l => ({
    type: 'Feature',
    id: l.id,
    geometry: {
      type: 'Point',
      coordinates: [l.longitude, l.latitude]
    },
    properties: { ...l }
  }))
}

export const mutations: MutationTree<RootState> = {
  ADD_LISTINGS (state, listings: any[]) {
    const propertyListings = listings
      .filter(l => l.type === 'PropertyListing')
      .map(({ listing }) => {
        const propertyListing: PropertyListing = {
          id: listing.id,
          price: listing.priceDetails?.displayPrice,
          area: listing.propertyDetails.area,
          bathrooms: listing.propertyDetails.bathrooms,
          bedrooms: listing.propertyDetails.bedrooms,
          carspaces: listing.propertyDetails.carspaces,
          displayableAddress: listing.propertyDetails.displayableAddress,
          features: listing.propertyDetails.features,
          landArea: listing.propertyDetails.landArea,
          latitude: listing.propertyDetails.latitude,
          longitude: listing.propertyDetails.longitude,
          postcode: listing.propertyDetails.postcode,
          propertyType: listing.propertyDetails.propertyType,
          region: listing.propertyDetails.region,
          state: listing.propertyDetails.state,
          street: listing.propertyDetails.street,
          streetNumber: listing.propertyDetails.streetNumber,
          suburb: listing.propertyDetails.suburb,
          unitNumber: listing.propertyDetails.unitNumber,
          media: listing.media.map((m: any) => m.url),
          listingSlug: listing.listingSlug
        }

        return propertyListing
      })

    state.list = propertyListings
  },
  UPDATE_PROPERTY_FILTER_TYPE (state, payload) {
    state.filter.properties.types = payload
  },
  UPDATE_PROPERTY_FILTER_ROOMS (state, payload) {
    state.filter.properties.rooms = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_BATHROOMS (state, payload) {
    state.filter.properties.bathrooms = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_CAR_SPACES (state, payload) {
    state.filter.properties.carSpaces = Number(payload)
  },
  UPDATE_PROPERTY_FILTER_MAX_PRICE (state, payload) {
    state.filter.properties.maxPrice = Number(payload)
  },
  UPDATE_BOUNDING_BOX (state, payload) {
    state.bounds.nw.lat = payload.getNorthWest().lat
    state.bounds.nw.lon = payload.getNorthWest().lng
    state.bounds.ne.lat = payload.getNorthEast().lat
    state.bounds.ne.lon = payload.getNorthEast().lng
    state.bounds.se.lat = payload.getSouthEast().lat
    state.bounds.se.lon = payload.getSouthEast().lng
    state.bounds.sw.lat = payload.getSouthWest().lat
    state.bounds.sw.lon = payload.getSouthWest().lng
  },
  UPDATE_SCHOOL_EDUCATION_SECTOR (state, payload) {
    state.filter.schools.educationSector = payload
  },
  UPDATE_SCHOOL_ZONE (state, payload) {
    state.filter.schools.zone = payload
  },
  UPDATE_PLOT_PRIMARY_SCHOOLS (state, payload) {
    state.filter.schools.primary.plot = payload
  },
  UPDATE_SCHOOL_PRIMARY_RATING (state, payload) {
    state.filter.schools.primary.rating = Number(payload)
  },
  UPDATE_SCHOOL_PRIMARY_ENG_RATING (state, payload) {
    state.filter.schools.primary.englishRating = Number(payload)
  },
  UPDATE_SCHOOL_PRIMARY_MATH_RATING (state, payload) {
    state.filter.schools.primary.mathsRating = Number(payload)
  },
  UPDATE_PLOT_SECONDARY_SCHOOLS (state, payload) {
    state.filter.schools.secondary.plot = payload
  },
  UPDATE_SCHOOL_SECONDARY_RATING (state, payload) {
    state.filter.schools.secondary.rating = Number(payload)
  },
  UPDATE_SCHOOL_SECONDARY_ENG_RATING (state, payload) {
    state.filter.schools.secondary.englishRating = Number(payload)
  },
  UPDATE_SCHOOL_SECONDARY_MATH_RATING (state, payload) {
    state.filter.schools.secondary.mathsRating = Number(payload)
  },
  ADD_GEOCODE_RESULTS (state, payload) {
    state.geocodeSuggestions = payload.map((f: any) => ({
      center: f.center,
      name: f.place_name,
      placeType: f.place_type
    }))
  }
}

export const actions: ActionTree<RootState, RootState> = {
  async loadListings ({ commit, state }) {
    const { filter, bounds } = state

    const payload = {
      listingType: 'Sale',
      pageSize: 200,
      propertyTypes: filter.properties.types,
      minBedrooms: filter.properties.rooms,
      minBathrooms: filter.properties.bathrooms,
      minCarspaces: filter.properties.carSpaces,
      maxPrice: filter.properties.maxPrice,
      locations: [{ state: 'VIC' }],
      geoWindow: { box: { topLeft: bounds.nw, bottomRight: bounds.se } }
    }

    const listings = await this.$axios.$post('https://api.domain.com.au/v1/listings/residential/_search', payload, { headers: {'X-Api-Key': process.env.domainApiKey } })
    commit('ADD_LISTINGS', listings)
  },

  async geocode ({ commit }, searchTerm) {
    if (searchTerm) {
      const params = {
        key: process.env.mapTilerSecret,
        bbox: '141.24847451171752,-39.624732230379166,148.68072548828115,-34.32302549095252',
        language: 'en'
      }

      const geocodeResults = await this.$axios.$get(`https://api.maptiler.com/geocoding/${searchTerm}.json`, { params })
      const schoolsResults = await this.$axios.$get(`//${process.env.domain}/schools/schools.json`)
      const mappedSchoolResults = schoolsResults.map((s: any) => ({
        center: [s.lon, s.lat],
        place_name: s.schoolName,
        place_type: 'school'
      })).filter((s: any) => s.place_name.toLowerCase().includes(searchTerm))

      commit('ADD_GEOCODE_RESULTS', [...geocodeResults.features, ...mappedSchoolResults])
    } else {
      commit('ADD_GEOCODE_RESULTS', [])
    }
  },

  setBoundingBox({ commit }, payload) {
    commit('UPDATE_BOUNDING_BOX', payload)
  }
}
