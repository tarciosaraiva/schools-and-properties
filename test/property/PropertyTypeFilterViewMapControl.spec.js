import { mount, createLocalVue } from '@vue/test-utils'
import { library } from '@fortawesome/fontawesome-svg-core'
import { fas } from '@fortawesome/free-solid-svg-icons'
import { FontAwesomeIcon } from '@fortawesome/vue-fontawesome'
import Vuex from 'vuex'

import { state, getters } from '@/store'
import PropertyTypeFilterViewMapControl from '@/components/property/PropertyTypeFilterViewMapControl.vue'

library.add(fas)

const localVue = createLocalVue()
localVue.use(Vuex)
localVue.component('FontAwesomeIcon', FontAwesomeIcon)

describe('PropertyTypeFilterViewMapControl', () => {
  let wrapper, store, actions, mutations, removeEventListenerSpy

  beforeEach(() => {
    actions = {
      loadListings: jest.fn()
    }

    mutations = {
      'UPDATE_PROPERTY_FILTER_TYPE': jest.fn(),
    }

    removeEventListenerSpy = jest.spyOn(document, 'removeEventListener')

    store = new Vuex.Store({ state, getters, actions, mutations })
    wrapper = mount(PropertyTypeFilterViewMapControl, { store, localVue })
  })

  test('is a Vue instance', () => {
    expect(wrapper.vm).toBeTruthy()
  })

  test('renders correctly when collapsed', () => {
    expect(wrapper).toMatchSnapshot()
  })

  test('renders correctly when expanded', async () => {
    await wrapper.setData({ collapsed: false })
    expect(wrapper).toMatchSnapshot()
  })

  test('should display control once bath button is clicked', async () => {
    let expandedControl = wrapper.find('ul.layer-options.expanded')
    expect(expandedControl.exists()).toBe(false)

    const buttons = wrapper.findAll('button')
    await buttons.at(0).trigger('click')

    expandedControl = wrapper.find('ul.layer-options.expanded')
    expect(expandedControl.exists()).toBe(true)
  })

  test('should trigger listing query once confirm button is clicked', async () => {
    await wrapper.setData({ collapsed: false })
    const buttons = wrapper.findAll('button')
    await buttons.at(1).trigger('click')

    expect(wrapper.vm.collapsed).toBe(true)
    expect(actions.loadListings).toHaveBeenCalled()
  })

  test('should have committed mutation to the store when value changes', async () => {
    await wrapper.setData({ propertyTypes: ['Unit'] })
    expect(mutations.UPDATE_PROPERTY_FILTER_TYPE).toHaveBeenCalledWith(store.state, ['Unit'])
  })

  test('should remove event listener when destroyed', () => {
    wrapper.destroy()
    expect(removeEventListenerSpy).toHaveBeenCalledWith('click', wrapper.vm.handleClickOutside)
  })
})
