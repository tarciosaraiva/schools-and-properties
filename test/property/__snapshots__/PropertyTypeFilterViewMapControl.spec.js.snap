// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PropertyTypeFilterViewMapControl renders correctly when collapsed 1`] = `
<div class="layer-container">
  <button title="Show property type filter control">
    <svg aria-hidden="true" class="svg-inline--fa fa-home fa-w-18" data-icon="home" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z" fill="currentColor" />
    </svg>
  </button>
  <button style="display: none;" title="Query listings with current property types">
    <svg aria-hidden="true" class="svg-inline--fa fa-check fa-w-16 fa-lg" data-icon="check" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z" fill="currentColor" />
    </svg>
  </button>
  <ul class="layer-options">
    <li>
      <label for="p_house_type">
        <input id="p_house_type" type="checkbox" value="House"> House
      </label>
    </li>
    <li>
      <label for="p_townhouse_type">
        <input id="p_townhouse_type" type="checkbox" value="Townhouse"> Townhouse
      </label>
    </li>
    <li>
      <label for="p_unit_type">
        <input id="p_unit_type" type="checkbox" value="ApartmentUnitFlat"> Unit
      </label>
    </li>
  </ul>
</div>
`;

exports[`PropertyTypeFilterViewMapControl renders correctly when expanded 1`] = `
<div class="layer-container">
  <button title="Show property type filter control">
    <svg aria-hidden="true" class="svg-inline--fa fa-home fa-w-18" data-icon="home" data-prefix="fas" focusable="false" role="img" viewBox="0 0 576 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M280.37 148.26L96 300.11V464a16 16 0 0 0 16 16l112.06-.29a16 16 0 0 0 15.92-16V368a16 16 0 0 1 16-16h64a16 16 0 0 1 16 16v95.64a16 16 0 0 0 16 16.05L464 480a16 16 0 0 0 16-16V300L295.67 148.26a12.19 12.19 0 0 0-15.3 0zM571.6 251.47L488 182.56V44.05a12 12 0 0 0-12-12h-56a12 12 0 0 0-12 12v72.61L318.47 43a48 48 0 0 0-61 0L4.34 251.47a12 12 0 0 0-1.6 16.9l25.5 31A12 12 0 0 0 45.15 301l235.22-193.74a12.19 12.19 0 0 1 15.3 0L530.9 301a12 12 0 0 0 16.9-1.6l25.5-31a12 12 0 0 0-1.7-16.93z" fill="currentColor" />
    </svg>
  </button>
  <button style title="Query listings with current property types">
    <svg aria-hidden="true" class="svg-inline--fa fa-check fa-w-16 fa-lg" data-icon="check" data-prefix="fas" focusable="false" role="img" viewBox="0 0 512 512" xmlns="http://www.w3.org/2000/svg">
      <path class="" d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z" fill="currentColor" />
    </svg>
  </button>
  <ul class="layer-options expanded">
    <li>
      <label for="p_house_type">
        <input id="p_house_type" type="checkbox" value="House"> House
      </label>
    </li>
    <li>
      <label for="p_townhouse_type">
        <input id="p_townhouse_type" type="checkbox" value="Townhouse"> Townhouse
      </label>
    </li>
    <li>
      <label for="p_unit_type">
        <input id="p_unit_type" type="checkbox" value="ApartmentUnitFlat"> Unit
      </label>
    </li>
  </ul>
</div>
`;
